### ISSUE TITLE
roslaunch: ignores ROS_NAMESPACE for parameters

### ISSUE THREAD

**Issue by 130s** [ros/ros_comm#58]
Migrated from https://code.ros.org/trac/ros/ticket/2380
Attachment available in code.ros.org?: No

---

It makes working with multiple robots under one master less convenient.

**Comment by meyerj** [comment]
Are there any news regarding this issue?

This bug breaks multiple robot support in Gazebo, at least without the need to modify the launch files used for a single robot:

```
<launch>
   <!-- send the robot XML to param server -->
   <param name="robot_description" command="$(find xacro)/xacro.py $(find hector_quadrotor_urdf)/urdf/quadrotor.gazebo.xacro" />

   <!-- push robot_description to factory and spawn robot in gazebo -->
   <node name="spawn_robot" pkg="gazebo_ros" type="spawn_model"
   args="-param robot_description
   -urdf
   -z 0.5
   -model quadrotor"
     respawn="false" output="screen"/>
</launch>

$ ROS_NAMESPACE=quadrotor1 roslaunch hector_quadrotor_gazebo spawn_quadrotor.launch
$ ROS_NAMESPACE=quadrotor2 roslaunch hector_quadrotor_gazebo spawn_quadrotor.launch
```

roslaunch sets the robot_description in the root namespace and ignores the ROS_NAMESPACE environment variable. The nodes itself are respecting ROS_NAMESPACE and therefore cannot find parameters set by roslaunch:

```
[INFO] [WallTime: 1373482991.569527] [0.000000] Loading model xml from ros parameter
Traceback (most recent call last):
  File "/home/meyer/ros/hydro/gazebo/src/gazebo_ros_pkgs/gazebo_ros/scripts/spawn_model", line 292, in <module>
    sm.callSpawnService()
  File "/home/meyer/ros/hydro/gazebo/src/gazebo_ros_pkgs/gazebo_ros/scripts/spawn_model", line 229, in callSpawnService
    model_xml = rospy.get_param(self.param_name)
  File "/opt/ros/hydro/lib/python2.7/dist-packages/rospy/client.py", line 452, in get_param
    return _param_server[param_name] #MasterProxy does all the magic for us
  File "/opt/ros/hydro/lib/python2.7/dist-packages/rospy/msproxy.py", line 117, in __getitem__
    raise KeyError(key)
KeyError: 'robot_description'
```

In the case of the `robot_description` parameters one could simply use `/robot_description` instead if all robots are the same, but in general there might be parameters that are different for each robot.

As a workaround, I am currently adding an enclosing `<group ns="$(optenv ROS_NAMESPACE)">...</group>` tag in all relevant launch files. This solution works fine and should IMHO reflect the builtin roslaunch semantics.

**Comment by dirk-thomas** [comment]
I have updated the issue title to mention that _roslaunch_ ignores the `ROS_NAMESPACE` environment variable when evaluating _param_ tags.

**Comment by dirk-thomas** [comment]
It currently also pushes nodes into a namespace even if the specify an absolute namespace themselves (since it does not reset the environment when invoking the node).

**Comment by dirk-thomas** [comment]
Last but not least ROS_NAMESPACE currently pushes the parameters and nodes from roscore.xml into a sub namespace which is also unintended.

**Comment by dirk-thomas** [comment]
@meyerj I have create a branch (https://github.com/ros/ros_comm/tree/roslaunch_ROS_NAMESPACE) which updates roslaunch to support ROS_NAMESPACE. Can you please give it a try and check if it works for you as expected?

**Comment by dirk-thomas** [comment]
If you give it a try you have to remove you global copy of the `roscore.xml` file from `/opt/ros/groovy/etc/ros/` else it will not use the modified file from the working copy.

**Comment by meyerj** [comment]
I tried to clone the roslaunch_ROS_NAMESPACE and to build it using catkin_make.
I got the following cmake error:

```
$ catkin_make --source ros_comm
Base path: /shared/home/meyer/ros/hydro
Source space: /shared/home/meyer/ros/hydro/ros_comm
Build space: /shared/home/meyer/ros/hydro/build
Devel space: /shared/home/meyer/ros/hydro/devel
Install space: /shared/home/meyer/ros/hydro/install
Creating symlink "/shared/home/meyer/ros/hydro/ros_comm/CMakeLists.txt" pointing to "/opt/ros/hydro/share/catkin/cmake/toplevel.cmake"
####
#### Running command: "cmake /shared/home/meyer/ros/hydro/ros_comm -DCATKIN_DEVEL_PREFIX=/shared/home/meyer/ros/hydro/devel -DCMAKE_INSTALL_PREFIX=/shared/home/meyer/ros/hydro/install" in "/shared/home/meyer/ros/hydro/build"
####
-- The C compiler identification is GNU
-- The CXX compiler identification is GNU
-- Check for working C compiler: /usr/lib/ccache/gcc
-- Check for working C compiler: /usr/lib/ccache/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/lib/ccache/c++
-- Check for working CXX compiler: /usr/lib/ccache/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Using CATKIN_DEVEL_PREFIX: /shared/home/meyer/ros/hydro/devel
-- Using CMAKE_PREFIX_PATH: /home/meyer/ros/hydro/gazebo/devel;/opt/ros/hydro;/opt/hector/hydro/devel;/opt/hector/hydro/devel
-- This workspace overlays: /home/meyer/ros/hydro/gazebo/devel;/opt/ros/hydro
-- Found PythonInterp: /usr/bin/python (found version "2.7.3")
-- Looking for include files CMAKE_HAVE_PTHREAD_H
-- Looking for include files CMAKE_HAVE_PTHREAD_H - found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE 
-- Found gtest sources under '/usr/src/gtest': gtests will be built
-- Using CATKIN_TEST_RESULTS_DIR: /shared/home/meyer/ros/hydro/build/test_results
-- catkin 0.5.67
-- BUILD_SHARED_LIBS is on
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-- ~~  traversing 30 packages in topological order:
-- ~~  - ros_comm (metapackage)
-- ~~  - rosconsole
-- ~~  - rosgraph
-- ~~  - rosgraph_msgs
-- ~~  - roslaunch
-- ~~  - rosmaster
-- ~~  - rosmsg
-- ~~  - rosparam
-- ~~  - rospy
-- ~~  - rosservice
-- ~~  - rostest
-- ~~  - std_srvs
-- ~~  - test_roslib_comm
-- ~~  - xmlrpcpp
-- ~~  - roscpp
-- ~~  - rosout
-- ~~  - message_filters
-- ~~  - rosnode
-- ~~  - rostopic
-- ~~  - roswtf
-- ~~  - test_roscpp
-- ~~  - test_rosgraph
-- ~~  - test_roslaunch
-- ~~  - test_rosmaster
-- ~~  - test_rosparam
-- ~~  - test_rospy
-- ~~  - test_rosservice
-- ~~  - topic_tools
-- ~~  - rosbag
-- ~~  - test_rosbag
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-- +++ processing catkin metapackage: 'ros_comm'
-- ==> add_subdirectory(ros_comm)
-- +++ processing catkin package: 'rosconsole'
-- ==> add_subdirectory(tools/rosconsole)
-- Boost version: 1.46.1
-- Found the following Boost libraries:
--   regex
--   thread
-- +++ processing catkin package: 'rosgraph'
-- ==> add_subdirectory(tools/rosgraph)
-- +++ processing catkin package: 'rosgraph_msgs'
-- ==> add_subdirectory(messages/rosgraph_msgs)
-- rosgraph_msgs: 2 messages, 0 services
-- +++ processing catkin package: 'roslaunch'
-- ==> add_subdirectory(tools/roslaunch)
-- +++ processing catkin package: 'rosmaster'
-- ==> add_subdirectory(tools/rosmaster)
-- +++ processing catkin package: 'rosmsg'
-- ==> add_subdirectory(tools/rosmsg)
-- +++ processing catkin package: 'rosparam'
-- ==> add_subdirectory(tools/rosparam)
-- +++ processing catkin package: 'rospy'
-- ==> add_subdirectory(clients/rospy)
-- +++ processing catkin package: 'rosservice'
-- ==> add_subdirectory(tools/rosservice)
-- +++ processing catkin package: 'rostest'
-- ==> add_subdirectory(tools/rostest)
-- +++ processing catkin package: 'std_srvs'
-- ==> add_subdirectory(messages/std_srvs)
-- std_srvs: 0 messages, 1 services
-- +++ processing catkin package: 'test_roslib_comm'
-- ==> add_subdirectory(test/test_roslib_comm)
-- +++ processing catkin package: 'xmlrpcpp'
-- ==> add_subdirectory(utilities/xmlrpcpp)
-- +++ processing catkin package: 'roscpp'
-- ==> add_subdirectory(clients/roscpp)
-- Boost version: 1.46.1
-- Found the following Boost libraries:
--   signals
--   filesystem
--   system
-- roscpp: 1 messages, 3 services
-- Looking for include files HAVE_IFADDRS_H
-- Looking for include files HAVE_IFADDRS_H - found
-- Looking for trunc
-- Looking for trunc - not found
-- +++ processing catkin package: 'rosout'
-- ==> add_subdirectory(tools/rosout)
-- +++ processing catkin package: 'message_filters'
-- ==> add_subdirectory(utilities/message_filters)
-- +++ processing catkin package: 'rosnode'
-- ==> add_subdirectory(tools/rosnode)
CMake Error at /shared/home/meyer/ros/hydro/devel/share/rostest/cmake/rostest-extras.cmake:3 (_generate_function_if_testing_is_disabled):
  Unknown CMake command "_generate_function_if_testing_is_disabled".
Call Stack (most recent call first):
  /shared/home/meyer/ros/hydro/devel/share/rostest/cmake/rostestConfig.cmake:173 (include)
  /opt/ros/hydro/share/catkin/cmake/catkinConfig.cmake:72 (find_package)
  tools/rosnode/CMakeLists.txt:3 (find_package)


-- Configuring incomplete, errors occurred!
Invoking "cmake" failed
```

**Comment by dirk-thomas** [comment]
You also need a current checkout of catkin in your workspace (see https://github.com/ros/ros_comm/blob/groovy-devel/tools/rosnode/package.xml#L17).

**Comment by meyerj** [comment]
Okay, thanks Dirk. Instead of building everything using catkin I tried to simply add the branched version of roslaunch to the ROS_PACKAGE_PATH and to add roslaunch/src to the PYTHONPATH variable. On first sight everything works as expected now (parameters are set in the correct namespace respecting ROS_NAMESPACE).

I also tried to add node tags that have an additional `ns=` attribute in the roslaunch XML itself and some private parameters. Relative namespaces are added to the namespace specified with ROS_NAMESPACE. Absolute namespace tags are used as is as expected. The parameters are also set in the correct namespaces now.

The only issue that remains is getting the tf tree and the robot model stuff right based on `tf_prefix`, but that's not within the scope of this issue.

Thanks, @dirk-thomas, for the quick response.
